nifi-nar-bundles/nifi-cdc/nifi-cdc-postgresql-bundle/nifi-cdc-postgresql-processors/src/main/java/org/apache/nifi/cdc/postgresql/processors/CaptureChangePostgreSQL.java

```diff
@@ -0,0 +1,468 @@
+        this.initialLSN = context.getProperty(INIT_LSN).asLong() == null ? null : context.getProperty(INIT_LSN).asLong();
+
+        try {
+
+            // Ensure driverLocation and driverName are correct before establishing
+            // PostgreSQL connection to avoid failing after PostgreSQL messages are received.
+            // Actual JDBC connection is created after PostgreSQL client gets started,
+            // because we need the connect-able host same as the PostgreSQL client.
+            registerDriver(driverLocation, driverName);
+            this.connectionManager = new ConnectionManager();
+            this.connectionManager.setProperties(host, database, username, password, driverName);
+            this.connectionManager.createSQLConnection();
+            this.connectionManager.createReplicationConnection();
+
+            stateMap = stateManager.getState(Scope.CLUSTER);
+
+            this.pgEasyReplication = new PGEasyReplication(publicationName, replicationSlotName, this.connectionManager);
+            this.pgEasyReplication.initializeLogicalReplication(dropSlotIfExists);
+
+            if (stateMap.get(LAST_LSN_RECEIVED) != null) {
+                this.lastLSNReceived = Long.parseLong(stateMap.get(LAST_LSN_RECEIVED));
+                this.hasRun.set(true);
+            } else {
+                this.hasRun.set(false);
+            }
+
+        } catch (final Exception ioe) {
+            logger.error("Failed to retrieve values from the State Manager. Will not attempt " + "connection until this is accomplished.", ioe);
+            context.yield();
+            throw new ProcessException(ioe.getMessage(), ioe);
+        }
+    }
+
+    List<Event> getData() {
+
+        List<Event> events = new ArrayList<>();
+
+        if (!this.hasRun.get() && this.hasSnapshot) {
+            events.add(this.pgEasyReplication.getSnapshot());
+        }
+
+        if (!this.hasRun.get() && this.initialLSN != null) {
+            events.add(this.pgEasyReplication.readEvent(true, this.includeBeginCommit, MIME_TYPE_VALUE, initialLSN));
+        }
+
+        events.add(this.pgEasyReplication.readEvent(true, this.includeBeginCommit));
+
+        if (events != null)
+            this.hasRun.set(true);
+
+        return events;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final ComponentLog logger = getLogger();
+
+        // Indicate that this processor has executed at least once, so we know whether
+        // or not the state values are valid and should be updated
+        if (this.pgEasyReplication == null) {
+            setup(context);
+        }
+
+        List<FlowFile> listFlowFiles = new ArrayList<>();
+
+        List<Event> events = getData();
+
+        if (events == null) {
+            logger.error("Failed to retrieve events.");
+            context.yield();
+            throw new ProcessException();
```

Comment:
Probably should include a helpful error message in the ProcessException as that's what will be displayed in the bulletin.

Reply:
Done!